File: ios/Sources/FlowbizOnsite/EventSerializer.swift
import Foundation

/// Event → wire serialization (SPEC §5).
///
/// Maps a typed ``Event`` to its wire event name and its `data` payload JSON
/// string: snake_case keys, optional (`nil`) fields omitted entirely.
/// Free-form `properties` / `recoveryProperties` maps are passed through with
/// their keys untouched; `.null` values inside objects are dropped (inside
/// arrays a `.null` element is kept as JSON `null` to preserve positions).
///
/// Must stay behaviorally identical to the Kotlin `EventSerializer` — both
/// are pinned by the shared fixtures in `shared/fixtures/`. The produced wire
/// string is rendered by ``CanonicalJSON`` (sorted keys, `JSON.stringify`
/// number rendering and escaping) and is byte-identical across platforms.
///
/// Garbage-input contract (aligned with Kotlin): serialization **throws** on
/// non-finite numbers (NaN/±Infinity) — org.json throws `JSONException` for
/// the same input. SPEC §3's never-throw boundary is applied at the public
/// API in Slice 4; internally serialization is strict.
enum EventSerializer {

    /// Wire event name (SPEC §5 table).
    static func wireName(_ event: Event) -> String {
        switch event {
        case .pageView: return "page.view"
        case .accountLogin: return "account.login"
        case .accountSync: return "account.sync"
        case .productView: return "product.view"
        case .cartSync: return "cart.sync"
        case .addToCart: return "cart.add"
        case .cartItemUpdate: return "cart.item.update"
        case .cartSetPostalCode: return "cart.setpostalcode"
        case .cartSetCoupon: return "cart.setcoupon"
        case .checkoutStep: return "checkout.step"
        case .orderComplete: return "order.complete"
        case .orderCancel: return "order.cancel"
        }
    }

    /// The envelope `data` field value: the payload as a canonical JSON
    /// string (see ``CanonicalJSON``). Throws only for non-finite numbers.
    static func dataJSONString(_ event: Event, baseUri: String? = nil) throws -> String {
        try CanonicalJSON.render(dataObject(event, baseUri: baseUri))
    }

    /// The payload as a JSON-compatible dictionary (snake_case keys, nils omitted).
    static func dataObject(_ event: Event, baseUri: String? = nil) -> [String: Any] {
        switch event {
        case .pageView(let path, let title):
            var page = [String: Any]()
            setIfPresent(&page, "title", title)
            setIfPresent(&page, "url", UrlResolver.resolve(path, baseUri: baseUri))
            return ["page": page]

        case .accountLogin(let user), .accountSync(let user):
            return ["user": userObject(user)]

        case .productView(let product):
            return ["product": productObject(product, baseUri: baseUri)]

        case .cartSync(let cart):
            return ["cart": cartObject(cart, baseUri: baseUri)]

        case .addToCart(let products):
            return ["products": products.map { cartItemObject($0, baseUri: baseUri) }]

        case .cartItemUpdate(let cartId, let productId, let sku, let quantity):
            return [
                "cart_id": cartId,
                "product_id": productId,
                "sku": sku,
                "quantity": quantity,
            ]

        case .cartSetPostalCode(let cartId, let postalCode):
            return ["cart_id": cartId, "postal_code": postalCode]

        case .cartSetCoupon(let cartId, let coupon):
            return ["cart_id": cartId, "coupon": coupon]

        case .checkoutStep(let checkout):
            return [
                "checkout": [
                    "cart_id": checkout.cartId,
                    "step": checkout.step,
                    "total_steps": checkout.totalSteps,
                    "step_name": checkout.stepName,
                ] as [String: Any]
            ]

        case .orderComplete(let order):
            return ["order": orderObject(order, baseUri: baseUri)]

        case .orderCancel(let orderId, let cartId):
            var payload = [String: Any]()
            setIfPresent(&payload, "order_id", orderId)
            setIfPresent(&payload, "cart_id", cartId)
            return payload
        }
    }

    private static func userObject(_ user: User) -> [String: Any] {
        var object: [String: Any] = [
            "user_id": user.userId,
            "email": user.email,
        ]
        setIfPresent(&object, "phone", user.phone)
        setIfPresent(&object, "name", user.name)
        setIfPresent(&object, "plan", user.plan)
        setIfPresent(&object, "created_at", user.createdAt)
        return object
    }

    private static func productObject(_ product: Product, baseUri: String?) -> [String: Any] {
        var object: [String: Any] = ["product_id": product.productId]
        setIfPresent(&object, "url", UrlResolver.resolve(product.url, baseUri: baseUri))
        setIfPresent(&object, "category", product.category)
        setIfPresent(&object, "brand", product.brand)
        object["variants"] = product.variants.map { variantObject($0, baseUri: baseUri) }
        return object
    }

    private static func variantObject(_ variant: ProductVariant, baseUri: String?) -> [String: Any] {
        var object: [String: Any] = [
            "sku": variant.sku,
            "price": variant.price,
        ]
        setIfPresent(&object, "name", variant.name)
        setIfPresent(&object, "url", UrlResolver.resolve(variant.url, baseUri: baseUri))
        setIfPresent(&object, "image_url", UrlResolver.resolve(variant.imageUrl, baseUri: baseUri))
        setIfPresent(&object, "price_from", variant.priceFrom)
        setIfPresent(&object, "stock", variant.stock)
        setIfPresent(&object, "available", variant.available)
        setIfPresent(&object, "properties", variant.properties.map(freeFormObject))
        setIfPresent(&object, "recovery_properties", variant.recoveryProperties.map(freeFormObject))
        return object
    }

    private static func cartItemObject(_ item: CartItem, baseUri: String?) -> [String: Any] {
        var object: [String: Any] = [
            "product_id": item.productId,
            "sku": item.sku,
            "quantity": item.quantity,
            "price": item.price,
        ]
        setIfPresent(&object, "name", item.name)
        setIfPresent(&object, "price_from", item.priceFrom)
        setIfPresent(&object, "category", item.category)
        setIfPresent(&object, "brand", item.brand)
        setIfPresent(&object, "url", UrlResolver.resolve(item.url, baseUri: baseUri))
        setIfPresent(&object, "image_url", UrlResolver.resolve(item.imageUrl, baseUri: baseUri))
        setIfPresent(&object, "properties", item.properties.map(freeFormObject))
        setIfPresent(&object, "recovery_properties", item.recoveryProperties.map(freeFormObject))
        return object
    }

    private static func addressObject(_ address: Address) -> [String: Any] {
        var object = [String: Any]()
        setIfPresent(&object, "postal_code", address.postalCode)
        setIfPresent(&object, "address_line1", address.addressLine1)
        setIfPresent(&object, "address_number", address.addressNumber)
        setIfPresent(&object, "address_line2", address.addressLine2)
        setIfPresent(&object, "city", address.city)
        setIfPresent(&object, "state", address.state)
        setIfPresent(&object, "country", address.country)
        setIfPresent(&object, "neighborhood", address.neighborhood)
        return object
    }

    private static func cartObject(_ cart: Cart, baseUri: String?) -> [String: Any] {
        var object: [String: Any] = [
            "cart_id": cart.cartId,
            "subtotal": cart.subtotal,
            "total": cart.total,
            "freight": cart.freight,
            "tax": cart.tax,
            "discounts": cart.discounts,
        ]
        setIfPresent(&object, "currency", cart.currency)
        setIfPresent(&object, "coupons", cart.coupons)
        setIfPresent(&object, "items", cart.items.map { $0.map { cartItemObject($0, baseUri: baseUri) } })
        setIfPresent(&object, "delivery_address", cart.deliveryAddress.map(addressObject))
        return object
    }

    private static func orderObject(_ order: Order, baseUri: String?) -> [String: Any] {
        var object: [String: Any] = ["cart_id": order.cartId]
        setIfPresent(&object, "order_id", order.orderId)
        object["subtotal"] = order.subtotal
        object["total"] = order.total
        object["freight"] = order.freight
        object["tax"] = order.tax
        object["discounts"] = order.discounts
        setIfPresent(&object, "currency", order.currency)
        setIfPresent(&object, "coupons", order.coupons)
        setIfPresent(&object, "items", order.items.map { $0.map { cartItemObject($0, baseUri: baseUri) } })
        setIfPresent(&object, "delivery_address", order.deliveryAddress.map(addressObject))
        setIfPresent(&object, "payment_methods", order.paymentMethods.map { methods in
            methods.map { ["type": $0.type, "amount": $0.amount] as [String: Any] }
        })
        setIfPresent(&object, "delivery_methods", order.deliveryMethods.map { methods in
            methods.map { ["type": $0.type, "amount": $0.amount] as [String: Any] }
        })
        return object
    }

    /// Sets `value` only when non-nil; nils are omitted entirely from the wire.
    private static func setIfPresent(_ object: inout [String: Any], _ key: String, _ value: Any?) {
        if let value { object[key] = value }
    }

    /// Free-form `properties` / `recovery_properties`: keys untouched, `.null` entries dropped.
    private static func freeFormObject(_ map: [String: JSONValue]) -> [String: Any] {
        JSONValue.object(map).foundationValue as? [String: Any] ?? [:]
    }
}

File: android/sdk/src/main/kotlin/com/flowbiz/onsite/EventSerializer.kt
package com.flowbiz.onsite

import org.json.JSONArray
import org.json.JSONObject

/**
 * Event → wire serialization (SPEC §5).
 *
 * Maps a typed [Event] to its wire event name and its `data` payload JSON
 * string: snake_case keys, optional (`null`) fields omitted entirely.
 * Free-form `properties` / `recoveryProperties` maps are passed through with
 * their keys untouched; `null` values inside those maps are dropped (inside
 * arrays a `null` element is kept as JSON `null` to preserve positions).
 *
 * Must stay behaviorally identical to the Swift `EventSerializer` — both are
 * pinned by the shared fixtures in `shared/fixtures/`. The produced wire
 * string is rendered by [CanonicalJson] (sorted keys, `JSON.stringify`
 * number rendering and escaping) and is byte-identical across platforms.
 *
 * Garbage-input contract (aligned with Swift): serialization **throws** on
 * non-finite numbers (NaN/±Infinity) — org.json throws `JSONException` when
 * the value enters the tree; the Swift writer throws its own error for the
 * same input. SPEC §3's never-throw boundary is applied at the public API in
 * Slice 4; internally serialization is strict.
 */
internal object EventSerializer {

    /** Wire event name (SPEC §5 table). */
    fun wireName(event: Event): String = when (event) {
        is Event.PageView -> "page.view"
        is Event.AccountLogin -> "account.login"
        is Event.AccountSync -> "account.sync"
        is Event.ProductView -> "product.view"
        is Event.CartSync -> "cart.sync"
        is Event.AddToCart -> "cart.add"
        is Event.CartItemUpdate -> "cart.item.update"
        is Event.CartSetPostalCode -> "cart.setpostalcode"
        is Event.CartSetCoupon -> "cart.setcoupon"
        is Event.CheckoutStep -> "checkout.step"
        is Event.OrderComplete -> "order.complete"
        is Event.OrderCancel -> "order.cancel"
    }

    /**
     * The envelope `data` field value: the payload as a canonical JSON string
     * (see [CanonicalJson]). Throws only for non-finite numbers.
     */
    fun dataJson(event: Event, baseUri: String? = null): String = CanonicalJson.render(dataObject(event, baseUri))

    /** The payload as a JSON object (snake_case keys, nulls omitted). */
    fun dataObject(event: Event, baseUri: String? = null): JSONObject = when (event) {
        is Event.PageView -> JSONObject().put("page", pageObject(event, baseUri))
        is Event.AccountLogin -> JSONObject().put("user", userObject(event.user))
        is Event.AccountSync -> JSONObject().put("user", userObject(event.user))
        is Event.ProductView -> JSONObject().put("product", productObject(event.product, baseUri))
        is Event.CartSync -> JSONObject().put("cart", cartObject(event.cart, baseUri))
        is Event.AddToCart -> JSONObject().put("products", JSONArray(event.products.map { cartItemObject(it, baseUri) }))
        is Event.CartItemUpdate -> JSONObject()
            .put("cart_id", event.cartId)
            .put("product_id", event.productId)
            .put("sku", event.sku)
            .put("quantity", event.quantity)
        is Event.CartSetPostalCode -> JSONObject()
            .put("cart_id", event.cartId)
            .put("postal_code", event.postalCode)
        is Event.CartSetCoupon -> JSONObject()
            .put("cart_id", event.cartId)
            .put("coupon", event.coupon)
        is Event.CheckoutStep -> JSONObject().put("checkout", checkoutObject(event.checkout))
        is Event.OrderComplete -> JSONObject().put("order", orderObject(event.order, baseUri))
        is Event.OrderCancel -> JSONObject()
            .putIfPresent("order_id", event.orderId)
            .putIfPresent("cart_id", event.cartId)
    }

    private fun pageObject(event: Event.PageView, baseUri: String?): JSONObject = JSONObject()
        .putIfPresent("title", event.title)
        .putIfPresent("url", UrlResolver.resolve(event.path, baseUri))

    private fun userObject(user: User): JSONObject = JSONObject()
        .put("user_id", user.userId)
        .put("email", user.email)
        .putIfPresent("phone", user.phone)
        .putIfPresent("name", user.name)
        .putIfPresent("plan", user.plan)
        .putIfPresent("created_at", user.createdAt)

    private fun productObject(product: Product, baseUri: String?): JSONObject = JSONObject()
        .put("product_id", product.productId)
        .putIfPresent("url", UrlResolver.resolve(product.url, baseUri))
        .putIfPresent("category", product.category)
        .putIfPresent("brand", product.brand)
        .put("variants", JSONArray(product.variants.map { variantObject(it, baseUri) }))

    private fun variantObject(variant: ProductVariant, baseUri: String?): JSONObject = JSONObject()
        .put("sku", variant.sku)
        .put("price", variant.price)
        .putIfPresent("name", variant.name)
        .putIfPresent("url", UrlResolver.resolve(variant.url, baseUri))
        .putIfPresent("image_url", UrlResolver.resolve(variant.imageUrl, baseUri))
        .putIfPresent("price_from", variant.priceFrom)
        .putIfPresent("stock", variant.stock)
        .putIfPresent("available", variant.available)
        .putIfPresent("properties", variant.properties?.let { freeFormObject(it) })
        .putIfPresent("recovery_properties", variant.recoveryProperties?.let { freeFormObject(it) })

    private fun cartItemObject(item: CartItem, baseUri: String?): JSONObject = JSONObject()
        .put("product_id", item.productId)
        .put("sku", item.sku)
        .put("quantity", item.quantity)
        .put("price", item.price)
        .putIfPresent("name", item.name)
        .putIfPresent("price_from", item.priceFrom)
        .putIfPresent("category", item.category)
        .putIfPresent("brand", item.brand)
        .putIfPresent("url", UrlResolver.resolve(item.url, baseUri))
        .putIfPresent("image_url", UrlResolver.resolve(item.imageUrl, baseUri))
        .putIfPresent("properties", item.properties?.let { freeFormObject(it) })
        .putIfPresent("recovery_properties", item.recoveryProperties?.let { freeFormObject(it) })

    private fun addressObject(address: Address): JSONObject = JSONObject()
        .putIfPresent("postal_code", address.postalCode)
        .putIfPresent("address_line1", address.addressLine1)
        .putIfPresent("address_number", address.addressNumber)
        .putIfPresent("address_line2", address.addressLine2)
        .putIfPresent("city", address.city)
        .putIfPresent("state", address.state)
        .putIfPresent("country", address.country)
        .putIfPresent("neighborhood", address.neighborhood)

    private fun cartObject(cart: Cart, baseUri: String?): JSONObject = JSONObject()
        .put("cart_id", cart.cartId)
        .put("subtotal", cart.subtotal)
        .put("total", cart.total)
        .put("freight", cart.freight)
        .put("tax", cart.tax)
        .put("discounts", cart.discounts)
        .putIfPresent("currency", cart.currency)
        .putIfPresent("coupons", cart.coupons?.let { JSONArray(it) })
        .putIfPresent("items", cart.items?.let { items -> JSONArray(items.map { cartItemObject(it, baseUri) }) })
        .putIfPresent("delivery_address", cart.deliveryAddress?.let { addressObject(it) })

    private fun checkoutObject(checkout: Checkout): JSONObject = JSONObject()
        .put("cart_id", checkout.cartId)
        .put("step", checkout.step)
        .put("total_steps", checkout.totalSteps)
        .put("step_name", checkout.stepName)

    private fun orderObject(order: Order, baseUri: String?): JSONObject = JSONObject()
        .put("cart_id", order.cartId)
        .putIfPresent("order_id", order.orderId)
        .put("subtotal", order.subtotal)
        .put("total", order.total)
        .put("freight", order.freight)
        .put("tax", order.tax)
        .put("discounts", order.discounts)
        .putIfPresent("currency", order.currency)
        .putIfPresent("coupons", order.coupons?.let { JSONArray(it) })
        .putIfPresent("items", order.items?.let { items -> JSONArray(items.map { cartItemObject(it, baseUri) }) })
        .putIfPresent("delivery_address", order.deliveryAddress?.let { addressObject(it) })
        .putIfPresent("payment_methods", order.paymentMethods?.let { methods ->
            JSONArray(methods.map { JSONObject().put("type", it.type).put("amount", it.amount) })
        })
        .putIfPresent("delivery_methods", order.deliveryMethods?.let { methods ->
            JSONArray(methods.map { JSONObject().put("type", it.type).put("amount", it.amount) })
        })

    /** Puts [value] only when non-null; nulls are omitted entirely from the wire. */
    private fun JSONObject.putIfPresent(key: String, value: Any?): JSONObject {
        if (value != null) put(key, wrapValue(value))
        return this
    }

    /** Free-form `properties` / `recovery_properties`: keys untouched, null entries dropped. */
    private fun freeFormObject(map: Map<*, *>): JSONObject {
        val obj = JSONObject()
        for ((key, value) in map) {
            if (key == null || value == null) continue
            obj.put(key.toString(), wrapValue(value))
        }
        return obj
    }

    /** Recursively converts Kotlin values to org.json values. */
    private fun wrapValue(value: Any): Any = when (value) {
        is JSONObject, is JSONArray -> value
        is Map<*, *> -> freeFormObject(value)
        is Collection<*> -> {
            val array = JSONArray()
            for (element in value) {
                array.put(if (element == null) JSONObject.NULL else wrapValue(element))
            }
            array
        }
        is String, is Boolean, is Number -> value
        else -> value.toString()
    }
}
